File: src/Scripting/PHP/example1.php
<?php

interface Button {
    public function render();
}

class DarkButton implements Button {
    public function render() {
        echo "Dark Button\n";
    }
}

class LightButton implements Button {
    public function render() {
        echo "Light Button\n";
    }
}

interface Checkbox {
    public function render();
}

class DarkCheckbox implements Checkbox {
    public function render() {
        echo "Dark Checkbox\n";
    }
}

class LightCheckbox implements Checkbox {
    public function render() {
        echo "Light Checkbox\n";
    }
}

interface UIFactory {
    public function createButton(): Button;
    public function createCheckbox(): Checkbox;
}

class DarkFactory implements UIFactory {
    public function createButton(): Button {
        return new DarkButton();
    }

    public function createCheckbox(): Checkbox {
        return new DarkCheckbox();
    }
}

class LightFactory implements UIFactory {
    public function createButton(): Button {
        return new LightButton();
    }

    public function createCheckbox(): Checkbox {
        return new LightCheckbox();
    }
}

function createUIComponents(UIFactory $factory) {
    $button = $factory->createButton();
    $checkbox = $factory->createCheckbox();
    $button->render();
    $checkbox->render();
}

// Usage
createUIComponents(new DarkFactory());
createUIComponents(new LightFactory());

File: src/Scripting/Perl/example1.pl
#!/usr/bin/perl
use strict;
use warnings;

# Concrete products
sub dark_button {
    print "Dark Button\n";
}

sub light_button {
    print "Light Button\n";
}

sub dark_checkbox {
    print "Dark Checkbox\n";
}

sub light_checkbox {
    print "Light Checkbox\n";
}

# Concrete factories. Each value represents one coherent product family.
sub dark_factory {
    return {
        create_button   => \&dark_button,
        create_checkbox => \&dark_checkbox,
    };
}

sub light_factory {
    return {
        create_button   => \&light_button,
        create_checkbox => \&light_checkbox,
    };
}

sub create_ui_components {
    my ($factory) = @_;
    $factory->{create_button}->();
    $factory->{create_checkbox}->();
}

# Usage: select the family once, then obtain every related product from it.
create_ui_components(dark_factory());
create_ui_components(light_factory());
